flang/docs/ArrayRepacking.md

```diff
@@ -0,0 +1,498 @@
+
+```Fortran
+program main
+  real :: x(2,10)
+  x = 0.0
+  call repacking_in(x(1,:))
+  call repacking_out(x(1,:))
+  print *, x
+contains
+  subroutine repacking_in(x)
+    real :: x(:)
+    !$acc enter data copyin(x)
+  end subroutine repacking_in
+  subroutine repacking_out(x)
+    real :: x(:)
+    !$acc exit data copyout(x)
+  end subroutine repacking_out
+end program main
+```
+
+So the most conservative implementation of the predicate generator may be to always produce `false` value. Flang lowering may attach any number of such attributes to `fir.pack_array` depending on the compilation context and options.
+
+[TBD] define `TempCopyIsSafe` attribute interface so that OpenACC/OpenMP dialects can provide their specific attributes, which can be used to generate static/runtime checks for safety of the temporary copy in particular context.
+
+### Lowering Fortran to FIR
+
+Fortran code:
+
+```Fortran
+subroutine test(x,n)
+  character(n) :: x(2:)
+end subroutine test
+```
+
+HLFIR/FIR produced by Lowering:
+
+```
+  func.func @_QPtest(%arg0: !fir.box<!fir.array<?x!fir.char<1,?>>> {fir.bindc_name = "x"}, %arg1: !fir.ref<i32> {fir.bindc_name = "n"}) {
+    %0 = fir.dummy_scope : !fir.dscope
+    %1:2 = hlfir.declare %arg1 dummy_scope %0 {uniq_name = "_QFtestEn"} : (!fir.ref<i32>, !fir.dscope) -> (!fir.ref<i32>, !fir.ref<i32>)
+    %2 = fir.load %1#0 : !fir.ref<i32>
+    %c0_i32 = arith.constant 0 : i32
+    %3 = arith.cmpi sgt, %2, %c0_i32 : i32
+    %4 = arith.select %3, %2, %c0_i32 : i32
+    %c2_i64 = arith.constant 2 : i64
+    %5 = fir.convert %c2_i64 : (i64) -> index
+    %6 = fir.shift %5 : (index) -> !fir.shift<1>
+    %repacked = fir.pack_array %arg0 typeparams %4 : !fir.box<!fir.array<?x!fir.char<1,?>>>
+    %7:2 = hlfir.declare %repacked(%6) typeparams %4 dummy_scope %0 {uniq_name = "_QFtestEx"} : (!fir.box<!fir.array<?x!fir.char<1,?>>>, !fir.shift<1>, i32, !fir.dscope) -> (!fir.box<!fir.array<?x!fir.char<1,?>>>, !fir.box<!fir.array<?x!fir.char<1,?>>>)
+    fir.unpack_array %repacked to %arg0 : !fir.box<!fir.array<?x!fir.char<1,?>>>
+    return
+  }
+```
+
+#### Restrictions:
+
+It is unsafe to create temporaries of assumed-shape dummy arrays that have `TARGET` attribute, because they can be accessed during the invocation of the subprograms not using direct reference of the dummy argument. Lowering must never produce the new operations for such dummy arguments. [TBD] a user option might be provided to override the default safe behavior.
+
+The copy creation is also restricted for `ASYNCHRONOUS` and `VOLATILE` arguments. Such dummy arguments might be changed during the execution of their subprogram in an unpredictable manner, so creating a copy for them might be incorrect (Fortran 2023, Note 5 of section 15.5.2.5).
```

Comment:
I wonder if openmp/acc programs that are broken by adding array packing/unpacking are really correct without using these attributes. Particularly in the case where the multithreaded element is in a different translation unit (or even subroutine/function).

Reply:
I think `Example #2` below is correct without ASYNCHRONOUS/VOLATILE. So there are cases where repacking breaks it.  Let me know if I misunderstood the comment, and you were talking about some other case.